File: autoparams/src/main/java/autoparams/generator/ConstructorResolverGenerator.java
package autoparams.generator;

import autoparams.ResolutionContext;

final class ConstructorResolverGenerator
    extends ObjectGeneratorBase<ConstructorResolver> {

    @Override
    protected ConstructorResolver generateObject(
        ObjectQuery query,
        ResolutionContext context
    ) {
        return generateResolver(context.resolve(ConstructorExtractor.class));
    }

    private static CompositeConstructorResolver generateResolver(
        ConstructorExtractor extractor
    ) {
        return new CompositeConstructorResolver(
            new AnnotatedModestConstructorResolver(extractor),
            new ModestConstructorResolver(extractor)
        );
    }
}

File: autoparams-kotlin-test/src/main/kotlin/autoparams/kotlin/ConstructorResolverGenerator.kt
package autoparams.kotlin

import autoparams.generator.ConstructorResolver
import autoparams.generator.ObjectContainer
import autoparams.generator.ObjectGenerator
import java.beans.ConstructorProperties
import java.util.Optional
import kotlin.reflect.full.primaryConstructor
import kotlin.reflect.jvm.javaConstructor

internal class ConstructorResolverGenerator : ObjectGenerator {

    private val value: ObjectContainer = ObjectContainer(
        ConstructorResolver
            .compose(
                ConstructorResolver { t: Class<*> ->
                    kotlin.runCatching { t.kotlin.primaryConstructor?.javaConstructor }
                        .getOrNull()
                        .let { Optional.ofNullable(it) }
                },
                ConstructorResolver { t: Class<*> ->
                    t.constructors
                        .filter { it.isAnnotationPresent(ConstructorProperties::class.java) }
                        .minByOrNull { it.parameterCount }
                        .let { Optional.ofNullable(it) }
                },
                ConstructorResolver { t: Class<*> ->
                    t.constructors
                        .minByOrNull { it.parameterCount }
                        .let { Optional.ofNullable(it) }
                }
            )
    )

    override fun generate(
        query: autoparams.generator.ObjectQuery,
        context: autoparams.generator.ObjectGenerationContext
    ): ObjectContainer {
        return if (query.type === ConstructorResolver::class.java)
            value
        else
            ObjectContainer.EMPTY
    }
}
